Pass text oid instead of invalid oid for null values

metdos opened the pull request:
Passing invalid oids even for null values in PQsendQueryParams() causes worker
nodes to fail. Therefore, we pass text oid for null values.

Fixes #752.

metdos commented:
Hey @anarazel, after discovering a bug in router executor selects, I have noticed a simpler solution to this problem and changed the code.

The first commit is a copy of PR 776 which is required for my fix. After PR 776 is merged to master, I will remove it from this PR.

anarazel commented on src/test/regress/sql/multi_prepare_plsql.sql:
```diff
@@ -247,6 +247,104 @@ SELECT plpgsql_test_2();
+END;
+$$ LANGUAGE plpgsql;
+
+SELECT no_parameter_insert();
```
Add comment along the lines of ```-- execute 6 times to trigger prepared statement usage ```

metdos replied:
> How about reducing error verbosity to ERROR?

That's a good idea, but still, job ids in the ERROR message can change because we run multi_prepare_sql and multi_prepare_plsql in parallel. We can run them sequentially, but the comment here notes that running them decreases regression test time 5 seconds.
https://github.com/citusdata/citus/issues/346#issue-134118154

What do you think? Do you have any preference?

anarazel commented on src/test/regress/sql/multi_prepare_plsql.sql:
```diff
@@ -247,6 +247,104 @@ SELECT plpgsql_test_2();
+SELECT partition_column_select(2);
+SELECT partition_column_select(3);
+SELECT partition_column_select(4);
+SELECT partition_column_select(5);
```
I'd rather test the failing case as well here, with a ```-- will fail``` comment

metdos replied:
That's a good point. This was my initial idea, but for failing case, the order of errors changes from execution to execution.

```sql
SELECT partition_column_select(6);
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57637
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57638
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57637
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57638
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57637
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57638
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57637
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57638
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57637
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
WARNING:  there is no parameter $1
CONTEXT:  while executing command on localhost:57638
PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
ERROR:  failed to execute job 780037
DETAIL:  Failure due to failed task 2
CONTEXT:  PL/pgSQL function partition_column_select(integer) line 4 at RETURN QUERY
```

anarazel commented on src/test/regress/sql/multi_prepare_plsql.sql:
```diff
@@ -247,6 +247,104 @@ SELECT plpgsql_test_2();
+SELECT non_partition_column_select(20);
+SELECT non_partition_column_select(30);
+SELECT non_partition_column_select(40);
+SELECT non_partition_column_select(50);
```
Why only five times here?

metdos replied:
This is similar to above comment. After 5th execution, we get errors in unpredictable orders.

anarazel commented:
This approach looks a lot better than the previous one...

anarazel commented:
On 2016-09-21 00:33:04 -0700, Metin Döşlü wrote:
> metdos commented on this pull request.
> 
> 
> 
> > +		temp_table.value
> +	FROM
> +		temp_table
> +	WHERE
> +		temp_table.key = key_arg
> +	ORDER BY
> +		key,
> +		value;
> +END;
> +$$ LANGUAGE plpgsql;
> +
> +SELECT partition_column_select(1);
> +SELECT partition_column_select(2);
> +SELECT partition_column_select(3);
> +SELECT partition_column_select(4);
> +SELECT partition_column_select(5);
> 
> That's a good point. This was my initial idea, but for failing case, the order of errors changes from execution to execution.

How about reducing error verbosity to ERROR?

metdos commented:
> How about reducing error verbosity to ERROR?

That's a good idea, but still, job ids in the ERROR message can change because we run multi_prepare_sql and multi_prepare_plsql in parallel. We can run them sequentially, but the comment here notes that running them decreases regression test time 5 seconds.
#346 (comment)

What do you think? Do you have any preference?

anarazel commented:
I think in that case it's ok to forgo the tests. But I suggest adding a comment about not wanting to run the failing test because of changing output.

metdos commented:
> I think in that case it's ok to forgo the tests. But I suggest adding a comment about not wanting to run the failing test because of changing output.

I updated comments and squashed commits related to this PR. After #776 is merged, I will drop the first commit, and merge this PR.

marcocitus commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -840,11 +841,14 @@ ExtractParametersFromParamListInfo(ParamListInfo paramListInfo, Oid **parameterT
 		if (parameterData->isnull || parameterData->ptype == 0)
 		{
 			(*parameterValues)[parameterIndex] = NULL;
+			(*parameterTypes)[parameterIndex] = TEXTOID;
```
This seems to cause #978
